llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp

```diff
@@ -2073,6 +2073,20 @@ Instruction *InstCombinerImpl::visitPtrToInt(PtrToIntInst &CI) {
     return InsertElementInst::Create(Vec, NewCast, Index);
   }
 
+  // (ptrtoint (gep (inttoptr Base), Offset)) -> Base + Offset
+  Value *Base, *Offset, *GEP;
+  if (match(SrcOp, m_OneUse(m_Value(GEP))) &&
+      match(GEP,
+            m_PtrAdd(m_OneUse(m_IntToPtr(m_Value(Base))), m_Value(Offset))) &&
```

Comment:
We could support this for arbitrary GEPs (not just ptradd) by calling `emitGEPOffset()` instead.